tests/test_candidate_routes.py

```diff
@@ -0,0 +1,103 @@
+import requests
+import setup
+import time
+
+from sqs_service import *
+import constants
+
+SQS_STAGE = "https://sqs.stage.osmosis.zone"
+
+ROUTES_URL = "/router/routes"
+
+# Test suite for the /router/routes endpoint
+class TestCandidateRoutes:
+    # Sanity check to ensure the test setup is correct
+    # before continunig with more complex test cases.
+    def test_usdc_uosmo(self, environment_url):
+        # Get max routes value from deployment config to expect the same number of candidate routes
+        # to be found
+        config = SQSService(environment_url).get_config()
+        expected_num_routes = config['Router']['MaxRoutes']
+        # Arbitrary choice based on performance at the time of test writing
+        expected_latency_upper_bound_ms = 300
+
+        self.run_candidate_routes_test(environment_url, constants.USDC, constants.UOSMO, expected_latency_upper_bound_ms, expected_min_routes=expected_num_routes, expected_max_routes=expected_num_routes)
+        
+    # Switch token in and out denoms compared to test_usdc_uosmo
+    def test_uosmo_usdc(self, environment_url):
+        # Get max routes value from deployment config to expect the same number of candidate routes
+        # to be found
+        config = SQSService(environment_url).get_config()
+        expected_num_routes = config['Router']['MaxRoutes']
+        # Arbitrary choice based on performance at the time of test writing
+        expected_latency_upper_bound_ms = 300
+
+        self.run_candidate_routes_test(environment_url, constants.UOSMO, constants.USDC, expected_latency_upper_bound_ms, expected_min_routes=expected_num_routes, expected_max_routes=expected_num_routes)
+
+    def run_candidate_routes_test(self, environment_url, token_in, token_out, expected_latency_upper_bound_ms, expected_min_routes, expected_max_routes):
+        """
+        Runs a test for the /router/routes endpoint with the given input parameters.
+
+        Returns routes for additional validation if needed by client
+
+        Validates:
+        - The number of routes returned
+        - Following pools in each route, all tokens within these pools are present and valid
+        - The latency is under the given bound
+        """
+        
+        sqs_service = SQSService(environment_url)
+
+        start_time = time.time()
+        response = sqs_service.get_candidate_routes(token_in, token_out)
+        elapsed_time_ms = (time.time() - start_time) * 1000
+
+        assert response.status_code == 200, f"Error: {response.text}"
+        assert expected_latency_upper_bound_ms > elapsed_time_ms, f"Error: latency {elapsed_time_ms} exceeded {expected_latency_upper_bound_ms} ms, token in {token_in} and token out {token_out}" 
```

Comment:
hmmm these latency tests could be unwieldy / hard to predict e.g

```
E       AssertionError: Error: latency 342.13829040527344 exceeded 300 ms, token in uosmo and token out ibc/D189335C6E4A68B513C10AB227BF1C1D38C746766278BA3EEB4FB14124F1D858
E       assert 300 > 342.13829040527344

tests/test_candidate_routes.py:56: AssertionError
```
||

```
E       AssertionError: Error: latency 363.1315231323242 exceeded 300 ms, token in ibc/D189335C6E4A68B513C10AB227BF1C1D38C746766278BA3EEB4FB14124F1D858 and token out uosmo
E       assert 300 > 363.1315231323242
```

Is there some way to improve these? Maybe upper and lower bounds, or something that wouldn't be flaky in e2e, can't really think of something rn, but thought I'd highlight here

Reply:
Also happy to merge and run in the CI to see if this is even an issue :thinking: 

but we should do that ASAP if that's the case

Reply:
300ms makes sense in production but build time may vary a lot depending on how busy the box is at that time.

Reply:
Yeah, that's a valid concern.

We will need to select a number that is high enough to not be flaky in CI while also catching regressions if any occur.

300ms was chosen as the right number for my local environment.

Once this is added to CI, I will work on increasing this to ~1s

There is no world where we would expect a query to take that long but if there is some obvious regression, this sanity check would be able to catch it quickly.